virtual_rainforest/core/grid.py

```diff
@@ -446,3 +452,132 @@ def populate_distances(
+
+        The function then returns a pair of arrays that give indices on the original
+        x and y data to extract data along a single cell id axis. If specific x and y
+        indexing data is not provided, they are assumed to be simple sequences along
+        `x_coords` and `y_coords`.
+
+        Args:
+            cell_map: An existing mapping from points onto cell ids.
+            strict: Should points outside the grid be allowed
+
+        Returns:
+            A list showing the integer cell id for each pair of points. If strict is
+            False, the list element for points outside the grid is None.
+        """
+
+        # Get the coordinate mapping to cell ids
+        cell_map = self.map_xy_to_cell_id(x_coords=x_coords, y_coords=y_coords)
+
+        # Set indexing to sequence along coords if missing
+        if (x_idx is None) and (y_idx is None):
+            _x_idx = np.arange(x_coords.shape[0])
+            _y_idx = np.arange(y_coords.shape[0])
+        elif (x_idx is None) ^ (y_idx is None):  # Note: ^ is xor
+            raise ValueError("Only one of x/y indices provided.")
+        else:
+            _x_idx = x_idx  # type: ignore
+            _y_idx = y_idx  # type: ignore
+
+        # Check provided
+        if (_x_idx.shape != x_coords.shape) or (  # type: ignore
+            _y_idx.shape != y_coords.shape  # type: ignore
+        ):  # type: ignore
+            raise ValueError("Dimensions of x/y indices do not match coordinates")
+
+        # Find the set of total number of cell mappings per point
+        cell_counts = set([len(mp) for mp in cell_map])
+
+        # If strict, raise an exception where not all coords fall in a grid cell
+        if strict and (0 in cell_counts):
+            raise ValueError("Mapped points fall outside grid.")
+
+        # Values greater than 1 indicate coordinates on cell edges
+        if any([c > 1 for c in cell_counts]):
+            raise ValueError("Mapped points fall on cell boundaries.")
+
+        # Now all points are 1 to 1 with cells, or missing if strict = False, so
+        # collapse down to list of ints/None.
+        cell_id_map = [c[0] if len(c) else None for c in cell_map]
+
+        # Get a list of all non-None ids.
+        cells_found = [v for v in cell_id_map if v is not None]
+
+        # Now check for cells with more than one point and cells with no points.
+        if set(cells_found) != set(self.cell_id):
+            raise ValueError("Mapped points do not cover all cells.")
+
+        if len(cells_found) != self.n_cells:
+            raise ValueError("Some cells contain more than one point.")
+
+        # Reduce to matching cells in cell_id order
+        cells = [
+            (mp, x, y)
+            for (mp, x, y) in zip(cell_id_map, _x_idx, _y_idx)
+            if mp is not None
+        ]
+        cells.sort()
+        _, x_idx, y_idx = zip(*cells)
+
+        return np.array(x_idx), np.array(y_idx)
```

Comment:
It is not clear to me what's happening here, are you mapping individual data points onto the grid or a full data set or is this part of the grid set up?

Reply:
Those last two functions are about mapping data on to the grid. 

* The `map_xy_to_cell_id` function checks pairs of xy points against any set of grid cell polygons (shape agnostic) and calculates which cells they occur in. For a given xy pair, can be:
    * `None` : point in no cells
    * E.g. `[1]`  point in cell 1.
    * E.g. `[1, 2, 3, 4]`  point on the corner of a cell and intersects _four_ cells.
* The other function `map_xy_to_cell_indexing` basically takes that and then does a load of checking to make sure that the set of points (from a dataset) uniquely maps onto and covers a given grid. The dataset can optionally cover a bigger region than the grid needs, as long as it can map in ok.